bioblend/galaxy/dataset_collections/__init__.py

```diff
@@ -80,3 +89,104 @@ def __init__(self, name, id):
+            'instance_type': instance_type,
+        }
+        url: str = self._make_url(module_id=dataset_collection_id)
+        return self._get(url=url, params=params)
+
+    def download_dataset_collection(self, dataset_collection_id: str, dir_path: Optional[str] = None,
+                                    use_default_dirname: bool = True, require_ok_state: bool = True,
+                                    maxwait: float = 12000) -> Optional[list]:
+        """
+        Download a dataset collection to memory or to disk.
+
+        When saving to memory, a list of dicts with be returned.
+        When saving to disk, a directory will be created containing the dataset collection.
+
+        If any dataset inside the collection does not have 'ok' states, a
+        ``DatasetStateException`` will be raised, unless ``require_ok_state=False``.
+
+        :type dataset_collection_id: str
+        :param dataset_collection_id: Encoded dataset collection ID
+
+        :type dir_path: str
+        :param dir_path: If this argument is no empty, the dataset will be streamed to disk
+          at that path (should be a directory if ``use_default_filename=True``).
+          If the file_path argument is not provided, the dataset content is loaded into memory
+          and returned by the method (Memory consumption may be heavy as the entire file
+          will be in memory).
+
+        :type use_default_dirname: bool
+        :param use_default_dirname: If ``True``, the export
+          directory will be named ``file_path/%s``,
+          where ``%s`` is the dataset collection name.
+          If ``False``, ``dir_path`` is assumed to
+          contain the full directory path including the directory name.
+
+        :type require_ok_state: bool
+        :param require_ok_state: If ``False``, datasets will be downloaded even if not in an 'ok' state,
+          issuing a ``DatasetStateWarning`` rather than raising a ``DatasetStateException``.
+
+        :type maxwait: float
+        :param maxwait: Total time (in seconds) to wait for the dataset state to
+          become terminal. If the dataset state is not terminal within this
+          time, a ``DatasetTimeoutException`` will be thrown.
+
+        :rtype: list of str
+        :return: If a ``dir_path`` argument is not provided, returns a list containing the dataset collection contents.
+          Otherwise returns nothing.
+        """
+        dataset_collection: dict = self.gi.dataset_collections.show_dataset_collection(dataset_collection_id)
+
+        if dir_path:
+            dir_path: str = os.path.join(dir_path, dataset_collection['name'] if use_default_dirname else '')
+            try:
+                os.mkdir(dir_path)
+            except OSError:
+                log.error(f'OSError: failed to create directory ``{dir_path}``')
+            else:
+                log.debug(f'Successfully created dataset collection download directory at ``{dir_path}``.')
+
```

Comment:
You could also just access `/api/dataset_collection/{hdca_id}/download` - unfortunately this endpoint doesn't seem very well documented.

Reply:
I did not see it among the API methods in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/dataset_collections.py, so I assumed it was not implemented.

Reply:
It is implemented, just in a different file: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/history_contents.py#L279, https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/buildapp.py#L541